From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Otto=20Pfl=C3=BCger?= <otto.pflueger@abscue.de>
Date: Sat, 1 Mar 2025 15:53:08 +0100
Subject: [PATCH 049/225] power: supply: sc27xx: force valid capacity range
 even if charging
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

If the capacity is less than 0 or greater than 100 due to a calibration
error, adjust it even if the battery is charging.

Signed-off-by: Otto Pflüger <otto.pflueger@abscue.de>
---
 drivers/power/supply/sc27xx_fuel_gauge.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/power/supply/sc27xx_fuel_gauge.c b/drivers/power/supply/sc27xx_fuel_gauge.c
index be6f3309c..356868253 100644
--- a/drivers/power/supply/sc27xx_fuel_gauge.c
+++ b/drivers/power/supply/sc27xx_fuel_gauge.c
@@ -839,10 +839,15 @@ static void sc27xx_fgu_capacity_calibration(struct sc27xx_fgu_data *data,
 
 	/*
 	 * If we are in charging mode, then we do not need to calibrate the
-	 * lower capacity.
+	 * lower capacity. Only adjust it if it is outside the valid range.
 	 */
-	if (chg_sts == POWER_SUPPLY_STATUS_CHARGING)
+	if (chg_sts == POWER_SUPPLY_STATUS_CHARGING) {
+		if (cap > 100)
+			sc27xx_fgu_adjust_cap(data, 100);
+		else if (cap < 0)
+			sc27xx_fgu_adjust_cap(data, 0);
 		return;
+	}
 
 	if ((ocv > data->cap_table[0].ocv && cap < 100) || cap > 100) {
 		/*
